Framework/DataHandling/src/LoadNXcanSAS.cpp

```diff
@@ -246,238 +281,171 @@ void loadInstrument(H5::Group &entry, const Mantid::API::MatrixWorkspace_sptr &w
+    }
   }
 
-  return hasQDev;
-}
-
-void loadData1D(H5::Group &dataGroup, const Mantid::API::MatrixWorkspace_sptr &workspace) {
-  // General
-  workspace->setDistribution(true);
-
-  // Load the Q value
-  workspace->mutableX(0) = Mantid::NeXus::H5Util::readArray1DCoerce<double>(dataGroup, sasDataQ);
-  workspace->getAxis(0)->setUnit("MomentumTransfer");
-
-  // Load the I value + units
-  workspace->mutableY(0) = Mantid::NeXus::H5Util::readArray1DCoerce<double>(dataGroup, sasDataI);
-
-  auto iDataSet = dataGroup.openDataSet(sasDataI);
-  auto yUnit = getUnit(iDataSet);
-  workspace->setYUnit(yUnit);
-
-  // Load the Idev value
-  workspace->mutableE(0) = Mantid::NeXus::H5Util::readArray1DCoerce<double>(dataGroup, sasDataIdev);
-
-  // Load the Qdev value (optional)
-  bool hasQResolution = hasQDev(dataGroup);
-  if (hasQResolution) {
-    workspace->setPointStandardDeviations(0, Mantid::NeXus::H5Util::readArray1DCoerce<double>(dataGroup, sasDataQdev));
+  void setUnits(const H5::DataSet &dataSet) const {
+    if (axisType == Y) {
+      workspace->setYUnit(getStrAttribute(dataSet, sasUnitAttr));
+    } else if (axisType == X) {
+      workspace->getAxis(0)->setUnit("MomentumTransfer");
+    }
   }
-}
-
-template <typename Functor>
-void read2DWorkspace(H5::DataSet &dataSet, Mantid::API::MatrixWorkspace_sptr workspace, Functor func,
-                     const H5::DataType &memoryDataType) {
-  using namespace Mantid::DataHandling::NXcanSAS;
-  auto dimInfo = getDataSpaceInfo(dataSet);
 
-  // File Data Space
-  auto fileSpace = dataSet.getSpace();
+  void setAxisType(const int type) { axisType = type; }
+  MatrixWorkspace_sptr workspace;
+  int axisType;
+};
 
-  // Size of single slab
-  const hsize_t rank = 2;
-  hsize_t sizeOfSingleSlab[rank] = {1, dimInfo.dimBin};
-  hsize_t start[rank] = {0, 0};
+/**
+ * Reads data from the h5 file and stores in the workspace through the data inserter struct
+ * @param dataSet: Dataset to read data from
+ * @param inserter: Inserter struct helping to insert data into the appropriate axes of the workspace
+ * @param slabShape: Vector containg the shape of the slice that's to be extracted from the dataset
+ * @param nPoints : Number of points (or bins) to select in the 1D slab.
+ * @param nHistograms: Number of histograms (iteration reference)
+ * @param offset: Offset position on the dataset.
+ */
+void readDataIntoWorkspace(const H5::DataSet &dataSet, const WorkspaceDataInserter &inserter,
+                           const std::vector<hsize_t> &slabShape, const hsize_t nPoints, const hsize_t nHistograms,
+                           std::vector<hsize_t> &offset) {
 
   // Memory Data Space
-  hsize_t memSpaceDimension[1] = {dimInfo.dimBin};
-  H5::DataSpace memSpace(1, memSpaceDimension);
+  const std::array<hsize_t, 1> memSpaceDimension = {nPoints};
+  const H5::DataSpace memSpace(1, memSpaceDimension.data());
+  const auto histogramIndex = slabShape.size() > 2 ? slabShape.size() - 2 : 0;
```

Comment:
This slab shape stuff is a little hard to follow, could we be explicit with what `slabShape.size() > 2` is representing?

`const bool isPolarisedDataset = slabShape.size() > 2` (I think!)